### PR TITLE
Fixed #35083 -- Make django.utils.decorators.method_decorator work with async functions.

### PR DESCRIPTION
**Pull request by VaarunSinha**
# Trac ticket number

[ticket-35083](https://code.djangoproject.com/ticket/35083)

# Branch description
I have checked if the function is a coroutine with the iscoroutine function, if it is then I have returned an async wrapper if not then I have returned a sync wrapper.

This solution does not introduce any new function or anything like that. 

Due to this the error which was coming before according to the ticket has disappeared 

# Checklist
- [x] This PR targets the `main` branch. 
- [x] The commit message is written in past tense, mentions the ticket number, and ends with a period.
- [x] I have checked the "Has patch" ticket flag in the Trac system.
- [x] I have added or updated relevant tests.
- [ ] I have added or updated relevant docs, including release notes if applicable.
- [ ] I have attached screenshots in both light and dark modes for any UI changes. - N/A

Please review the same and provide feedback!

### CODE REVIEW

**Comment by carltongibson** [comment]
Hi @VaarunSinha — you’ll need to add a regression test for this (so a test case showing method decorator in action on an async method — the example from the ticket is probably fine).

**Comment by VaarunSinha** [comment]
Got it will do.I am searching for similar tests so that I can figure out how to write them. Let me know if anyone has any.

File: django/utils/decorators.py [hunk 1]
```diff
@@ -15,8 +15,10 @@ def __get__(self, instance, cls=None):
 
 
 def _update_method_wrapper(_wrapper, decorator):
-    # _multi_decorate()'s bound_method isn't available in this scope. Cheat by
-    # using it on a dummy function.
+    """
+    Update the wrapper method to include the attributes from the decorator.
+    """
+
```
Comment (carltongibson) [inline django/utils/decorators.py:21]: I think you need to drop this (and other unrelated changes) to keep the patch minimal. 

If you think there's a case for these changes then they should be presented separately

File: django/utils/decorators.py [hunk 2]
```diff
@@ -30,42 +32,56 @@ def _multi_decorate(decorators, method):
     a single decorator or an iterable of decorators.
     """
     if hasattr(decorators, "__iter__"):
-        # Apply a list/tuple of decorators if 'decorators' is one. Decorator
-        # functions are applied so that the call order is the same as the
-        # order in which they appear in the iterable.
+        # If 'decorators' is a list or tuple, reverse it to apply in correct order.
         decorators = decorators[::-1]
     else:
+        # If it's a single decorator, make it a list.
```
Comment (carltongibson) [inline django/utils/decorators.py:38]: I think this comment just says what the code below it does no? (I don't think it adds anything)

File: django/utils/decorators.py [hunk 3]
```diff
@@ -30,42 +32,56 @@ def _multi_decorate(decorators, method):
...
 
-    def _wrapper(self, *args, **kwargs):
-        # bound_method has the signature that 'decorator' expects i.e. no
-        # 'self' argument, but it's a closure over self so it can call
-        # 'func'. Also, wrap method.__get__() in a function because new
-        # attributes can't be set on bound method objects, only on functions.
-        bound_method = wraps(method)(partial(method.__get__(self, type(self))))
+    if iscoroutinefunction(method):
+        # If the method is asynchronous, define an async wrapper.
+        async def _async_wrapper(self, *args, **kwargs):
```
Comment (carltongibson) [inline django/utils/decorators.py:43]: The variable can still be `_wrapper` in both branches I think.

**Review by carltongibson — COMMENTED** [review]
@VaarunSinha Looks on the right lines at first quick pass. 

Please revert the comment changes in the unchanged blocks, and try to minimise those in the changed block. (In particular, I think keeping the existing comments retains more useful information when defining the wrapper function. I'd want to mirror those in the async branch.)

File: django/utils/decorators.py (hunk 1 above)
Comment (VaarunSinha) [reply django/utils/decorators.py:21]: Got it I will drop this and other unrelated changes. For example the comments.

File: django/utils/decorators.py (hunk 3 above)
Comment (VaarunSinha) [reply django/utils/decorators.py:43]: The variable can be wrapper but I think making it sync and async helps in readability but will change if my point does not make sense.

File: django/utils/decorators.py [hunk 4]
```diff
@@ -37,35 +37,50 @@ def _multi_decorate(decorators, method):
...
 def method_decorator(decorator, name=""):
     """
-    Convert a function decorator into a method decorator
+    Convert a function decorator into a method decorator.
     """
 
-    # 'obj' can be a class or a function. If 'obj' is a function at the time it
-    # is passed to _dec,  it will eventually be a method of the class it is
-    # defined on. If 'obj' is a class, the 'name' is required to be the name
-    # of the method that will be decorated.
```
Comment (carltongibson) [inline django/utils/decorators.py:66]: Revert this.

File: django/utils/decorators.py [hunk 5]
```diff
@@ -37,35 +37,50 @@ def _multi_decorate(decorators, method):
...
+            _update_method_wrapper(_sync_wrapper, dec)
+        # Update the sync wrapper with the original method's attributes.
+        update_wrapper(_sync_wrapper, method)
+        return _sync_wrapper
 
 
 def method_decorator(decorator, name=""):
     """
-    Convert a function decorator into a method decorator
+    Convert a function decorator into a method decorator.
```
Comment (carltongibson) [inline django/utils/decorators.py:78]: Unrelated change.

File: django/utils/decorators.py [hunk 6]
```diff
@@ -37,35 +37,50 @@ def _multi_decorate(decorators, method):
...
     """
 
-    # 'obj' can be a class or a function. If 'obj' is a function at the time it
-    # is passed to _dec,  it will eventually be a method of the class it is
-    # defined on. If 'obj' is a class, the 'name' is required to be the name
-    # of the method that will be decorated.
     def _dec(obj):
         if not isinstance(obj, type):
+            # If 'obj' is a function, apply the decorator directly.
```
Comment (carltongibson) [inline django/utils/decorators.py:83]: Unrelated change.

File: django/utils/decorators.py [hunk 7]
```diff
@@ -78,15 +93,14 @@ def _dec(obj):
                 "Cannot decorate '%s' as it isn't a callable attribute of "
                 "%s (%s)." % (name, obj, method)
             )
+        # Apply the decorator to the method.
```
Comment (carltongibson) [inline django/utils/decorators.py:96]: Unrelated change.

File: django/utils/decorators.py [hunk 8]
```diff
@@ -78,15 +93,14 @@ def _dec(obj):
...
         _wrapper = _multi_decorate(decorator, method)
         setattr(obj, name, _wrapper)
         return obj
 
-    # Don't worry about making _dec look similar to a list/tuple as it's rather
-    # meaningless.
     if not hasattr(decorator, "__iter__"):
+        # Update the decorator function with attributes for debugging.
```
Comment (carltongibson) [inline django/utils/decorators.py:102]: unrelated change.

File: django/utils/decorators.py (hunk 3 above)
Comment (carltongibson) [reply django/utils/decorators.py]: Yes, please.

**Review by carltongibson — COMMENTED** [review]
Theres' still several unrelated changes. 

The block for the sync wrapper should be ≈exactly the exiting code (including comments) just indented a level. 

(Maybe that gets tweaked under review, but try to make the first pass as small a diff as possible.)

**Comment by VaarunSinha** [comment]
I have written tests and they are working also removed some more unneccesarry changes will squash into one commit once changes are approved for a cleaner merge.

File: django/utils/decorators.py [hunk 9]
```diff
@@ -37,33 +37,47 @@ def _multi_decorate(decorators, method):
...
+                bound_method = dec(bound_method)
+            # Call the decorated method.
+            return bound_method(*args, **kwargs)
+
+        # Update the sync wrapper with decorator attributes.
+        for dec in decorators:
+            _update_method_wrapper(_wrapper, dec)
+        # Update the sync wrapper with the original method's attributes.
+        update_wrapper(_wrapper, method)
+        return _wrapper
```
Comment (carltongibson) [inline django/utils/decorators.py:73]: You've still rewritten the comments from the existing code here. Please don't do that. 

Comments such as `# Call the decorated method.` followed by that exact code don't add anything.

**Review by carltongibson — COMMENTED** [review]
Hi @VaarunSinha. 

Well done for getting the test in place. 

* The existing tests for `method_decorator` are in `tests/decorators/tests.py` — we need to put the new tests there, to keep the related tests together. 
* Please add a release note to `docs/releases/5.2.txt` in the _Decorators_ section. 
* Please review my comment and reducing the diff, and not rewriting the comments in the existing code, and address that. 

Once that's done you can uncheck the _Patch needs improvement_, _Needs tests_, and _Needs documentation_ flags on the ticket-35083, and it will be inline for a review from a Fellow. 

Good work! 🚀

**Comment by VaarunSinha** [comment]
I have sqaushed all commits into one

**Comment by nessita** [comment]
> I have squashed all commits into one

Thank you for your work so far! Could you please ensure that you also follows this advice from Carlton?

_Once that's done you can uncheck the Patch needs improvement, Needs tests, and Needs documentation flags on the ticket-35083, and it will be inline for a review from a Fellow._

**Comment by VaarunSinha** [comment]
I merged all the commits from django main to here and in an attempt to rebase my branch to just contain one clean commit all other commits got pushed.

Is there any solution for this or do I have to delete my fork close this PR and create a new one?

@carltongibson @nessita

**Comment by carltongibson** [comment]
Hey @VaarunSinha — it looks like you've merged a branch from another PR here. Oops. These things happen... 

Likely the easiest thing to do is to re-apply your changes to a fresh branch. (There are other options but I'm not clear they're simpler to pickup.) 

So... 

1. Create a fresh branch from `main`, with no changes. 
2. Manually reapply the changes for this PR from your current branch. (Just the last commit, I think, so you can likely `cherrypick` it.) 
3. Re-commit. 
4. Set your remote branch for this PR as the tracking branch for the new local branch `git branch --set-upstream-to=...`
5. Force push to the remote branch to overwrite what you've got going on here: `git push --force`

That should set you right.

**Comment by VaarunSinha** [comment]
Closing this PR as this git issue is beyond repair will create a new PR and mention it here.
